bigquery_magics/graph_server.py

```diff
@@ -0,0 +1,248 @@
+# Copyright 2024 Google LLC
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     https://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import atexit
+import http.server
+import json
+import socketserver
+import threading
+from typing import Dict, List
+
+
+def convert_graph_data(query_results: Dict[str, Dict[str, str]]):
+    """
+    Converts graph data to the form expected by the visualization framework.
+
+    Receives graph data as a dictionary, produced by converting the underlying
+    DataFrame representing the query results into JSON, then into a
+    python dictionary. Converts it into a form expected by the visualization
+    framework.
+
+    Args:
+        query_results:
+            A dictionary with one key/value pair per column. For each column:
+            - The key is the name of the column (str)
+            - The value is another dictionary with one key/value pair per row.
+              Row each row:
+              - The key is a string that specifies the integer index of the row
+                (e.g. '0', '1', '2')
+              - The value is a JSON string containing the result of the query
+                for the current row/column. (Note: We only support graph
+                visualization for columns of type JSON).
+    """
+    # Delay spanner imports until this function is called to avoid making
+    # spanner-graph-notebook (and its dependencies) hard requirements for bigquery
+    # magics users, who don't need graph visualization.
+    #
+    # Note that these imports do not need to be in a try/except, as this function
+    # does not even get called unless spanner_graphs has already been confirmed
+    # to exist upstream.
+    from google.cloud.spanner_v1.types import StructType, Type, TypeCode
+    import networkx
+    from spanner_graphs.conversion import (
+        columns_to_native_numpy,
```

Comment:
As of https://github.com/cloudspannerecosystem/spanner-graph-notebook/blob/cf18014e5af6309bfee537d070278aaf76cc5d1b/spanner_graphs/conversion.py this method no longer exists.